File: built-in plugins dsp/mb-master-reference.cpp
/**
 * MB Reference Player
 * Category : effect
 * Type     : mastering
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : A/B reference comparison tool for mastering decisions
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MASTER_REFERENCE_H
#define MB_MASTER_REFERENCE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMasterReference : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-master-reference";
    static constexpr const char* PLUGIN_NAME    = "MB Reference Player";
    static constexpr const char* PLUGIN_TYPE    = "mastering";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float gain = 0f;  // range [-24, 24]
    float lowCut = 20f;  // range [20, 200]
    float highCut = 20000f;  // range [5000, 20000]
    };

    MbMasterReference() = default;
    ~MbMasterReference() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.gain = std::clamp(params.gain, -24f, 24f);
        params.lowCut = std::clamp(params.lowCut, 20f, 200f);
        params.highCut = std::clamp(params.highCut, 5000f, 20000f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Reference Player
        return input;
    }
};

#endif // MB_MASTER_REFERENCE_H

File: built-in plugins dsp/mb-master-reference.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbMasterReferencePlugin: PluginDefinition = { id: 'mb-master-reference', slug: 'mb-master-reference', name: 'MB Reference Player', category: 'effect', type: 'mastering' as any, version: '1.0.0', description: 'A/B reference comparison tool for mastering decisions', author: 'Max Booster', grade: 'A', parameters: [{ id: 'gain', name: 'Reference Gain', type: 'float', defaultValue: 0, minValue: -24, maxValue: 24, automatable: true }, { id: 'lowCut', name: 'Low Cut', type: 'float', defaultValue: 20, minValue: 20, maxValue: 200, automatable: true }, { id: 'highCut', name: 'High Cut', type: 'float', defaultValue: 20000, minValue: 5000, maxValue: 20000, automatable: true }], defaultPreset: { gain: 0, lowCut: 20, highCut: 20000 } };

export default MbMasterReferencePlugin;
